### ISSUE TITLE
Some confusion about analysis and experiment in the paper

### ISSUE THREAD

**Issue by theskyv** [ChaoningZhang/MobileSAM#13]
As I was reading the paper I had some confusion: your paper mentions that FastSAM needs at least two prompt points and on that basis compares the performance of FastSAM and MobileSAM in segment anything mode.

But as you can see from the [HuggingFace demo](https://huggingface.co/spaces/An-619/FastSAM) posted by FastSAM, their approach supports a single point and has worked well in my own attempts. If it is indeed a writing error, I very much look forward to your revision and showing new experiments to effectively compare the two methods.

BTW, I am also curious how Table 7 in the paper is done exactly, and why the model performance can be proved by modifying the distance of positive and negative prompt points?
![51db6846031ae9f807827f67fcbc1bc](https://github.com/ChaoningZhang/MobileSAM/assets/134121667/50f8b49f-4e84-4d12-9101-15245862e460)

**Comment by ChaoningZhang** [comment]
> As I was reading the paper I had some confusion: your paper mentions that FastSAM needs at least two prompt points and on that basis compares the performance of FastSAM and MobileSAM in segment anything mode.
> 
> But as you can see from the [HuggingFace demo](https://huggingface.co/spaces/An-619/FastSAM) posted by FastSAM, their approach supports a single point and has worked well in my own attempts. If it is indeed a writing error, I very much look forward to your revision and showing new experiments to effectively compare the two methods.
> 
> BTW, I am also curious how Table 7 in the paper is done exactly, and why the model performance can be proved by modifying the distance of positive and negative prompt points? ![51db6846031ae9f807827f67fcbc1bc](https://user-images.githubusercontent.com/134121667/249392317-50f8b49f-4e84-4d12-9101-15245862e460.png)

Thanks for your interest in our work. When we compared with their method and found that it cannot work with a single point. We will carefully read their code again to see how it works. 

For the results in Table7, we do not use the distance for proving that ours is better. Distance is more or less like a hyperparameter here and we show that FastSAM has a significantly lower with all distances. More extensive comparison is on the way.

**Comment by ChaoningZhang** [comment]
> As I was reading the paper I had some confusion: your paper mentions that FastSAM needs at least two prompt points and on that basis compares the performance of FastSAM and MobileSAM in segment anything mode.
> 
> But as you can see from the [HuggingFace demo](https://huggingface.co/spaces/An-619/FastSAM) posted by FastSAM, their approach supports a single point and has worked well in my own attempts. If it is indeed a writing error, I very much look forward to your revision and showing new experiments to effectively compare the two methods.
> 
> BTW, I am also curious how Table 7 in the paper is done exactly, and why the model performance can be proved by modifying the distance of positive and negative prompt points? ![51db6846031ae9f807827f67fcbc1bc](https://user-images.githubusercontent.com/134121667/249392317-50f8b49f-4e84-4d12-9101-15245862e460.png)

**Comment by ChaoningZhang** [comment]
> 

As stated in the paper, the mIoU was calculated with the comparsion with the mask generated by the original SAM. We set the point in the middle of the image. I suggested you to experiment with the mask with the original SAM. The mask you provided seems to be very good actually. I mean, if you try with the original SAM, you are very likely to get the same mask as the you reported. Detect small objects (window in this case) seems to be more preferable.

**Comment by theskyv** [comment]
Thank you for your reply and discussion. I have three new questions I would like to discuss with you @ChaoningZhang :

1. Is it important to output like with SAM or to output real and valid objects? -- As my pictures above show, FastSAM outputs a complete bus, while MobileSAM outputs a bus window with a lot of noise (>30%), and perhaps MobileSAM is more similar to SAM in mIoU. **However, I don't think "windows with a lot of noise" is more useful than "a complete bus"**.

2. Is the mIoU comparison method fair? --MobileSAM uses knowledge distillation to learn the output of the original SAM, while FastSAM is trained separately. It is well known that knowledge distillation helps make student more similar to teacher. **It feels very unfair** to take the original SAM as Ground Truth and compare the output of FastSAM and MobileSAM to see which is more similar to the original SAM. **This seems similar to "testing on training data"**.

3. Is FastSAM really not suitable for small objects? -- I tried looking at the segmentation results of FastSAM and found that it can segment car windows well (it's just that when faced with the segmentation ambiguity problem, FastSAM tends to prioritize large objects, while SAM/MobileSAM tends to prioritize small objects. This is not superior or inferior). Using the quantitative data in the FastSAM paper as evidence, **FastSAM performs due to SAM (ViT-H) on small objects and slightly inferior to SAM on large objects**. I'd like to know why you think FastSAM is not suitable for detecting small objects.

![image](https://github.com/ChaoningZhang/MobileSAM/assets/134121667/a10a1a84-321e-484a-ae48-7cd8897190cc)
![image](https://github.com/ChaoningZhang/MobileSAM/assets/134121667/02bc6efc-d910-479a-8405-9ee96531c8ee)


Looking forward to discuss with you.

**Comment by ChaoningZhang** [comment]
> Thank you for your reply and discussion. I have three new questions I would like to discuss with you @ChaoningZhang :
> 
> 1. Is it important to output like with SAM or to output real and valid objects? -- As my pictures above show, FastSAM outputs a complete bus, while MobileSAM outputs a bus window with a lot of noise (>30%), and perhaps MobileSAM is more similar to SAM in mIoU. **However, I don't think "windows with a lot of noise" is more useful than "a complete bus"**.
> 2. Is the mIoU comparison method fair? --MobileSAM uses knowledge distillation to learn the output of the original SAM, while FastSAM is trained separately. It is well known that knowledge distillation helps make student more similar to teacher. **It feels very unfair** to take the original SAM as Ground Truth and compare the output of FastSAM and MobileSAM to see which is more similar to the original SAM. **This seems similar to "testing on training data"**.
> 3. Is FastSAM really not suitable for small objects? -- I tried looking at the segmentation results of FastSAM and found that it can segment car windows well (it's just that when faced with the segmentation ambiguity problem, FastSAM tends to prioritize large objects, while SAM/MobileSAM tends to prioritize small objects. This is not superior or inferior). Using the quantitative data in the FastSAM paper as evidence, **FastSAM performs due to SAM (ViT-H) on small objects and slightly inferior to SAM on large objects**. I'd like to know why you think FastSAM is not suitable for detecting small objects.
> 
> ![image](https://user-images.githubusercontent.com/134121667/249666682-a10a1a84-321e-484a-ae48-7cd8897190cc.png) ![image](https://user-images.githubusercontent.com/134121667/249666700-02bc6efc-d910-479a-8405-9ee96531c8ee.png)
> 
> Looking forward to discuss with you.

We are currently very busy with other ongoing works and try to answer your question in a brief manner. An important characteristic of the SAM is that it can address the ambiguity issue. With a single point, SAM can output three masks to address these issues, for which our MobileSAM follows. If a large object is more meaningful than the smaller one as you suggested, SAM can just choose the three masks that have the largest region. Obviously, this does not make sense and that's why SAM relies on a score ranking mechanism to choose the final mask. In other words, it is the model that automatically prefers the smaller object (without human bias). It seems that FastSAM ignores this ambiguity issue. It seems to me that ignoring this ambiguity issue and claiming a larger object like bus might be biased. After all, for human A, a point can mean the bus, while for human B, the same point can mean the window in this context. Do you agree?

**Comment by ChaoningZhang** [comment]
> > Thank you for your reply and discussion. I have three new questions I would like to discuss with you @ChaoningZhang :
> > 
> > 1. Is it important to output like with SAM or to output real and valid objects? -- As my pictures above show, FastSAM outputs a complete bus, while MobileSAM outputs a bus window with a lot of noise (>30%), and perhaps MobileSAM is more similar to SAM in mIoU. **However, I don't think "windows with a lot of noise" is more useful than "a complete bus"**.
> > 2. Is the mIoU comparison method fair? --MobileSAM uses knowledge distillation to learn the output of the original SAM, while FastSAM is trained separately. It is well known that knowledge distillation helps make student more similar to teacher. **It feels very unfair** to take the original SAM as Ground Truth and compare the output of FastSAM and MobileSAM to see which is more similar to the original SAM. **This seems similar to "testing on training data"**.
> > 3. Is FastSAM really not suitable for small objects? -- I tried looking at the segmentation results of FastSAM and found that it can segment car windows well (it's just that when faced with the segmentation ambiguity problem, FastSAM tends to prioritize large objects, while SAM/MobileSAM tends to prioritize small objects. This is not superior or inferior). Using the quantitative data in the FastSAM paper as evidence, **FastSAM performs due to SAM (ViT-H) on small objects and slightly inferior to SAM on large objects**. I'd like to know why you think FastSAM is not suitable for detecting small objects.
> > 
> > ![image](https://user-images.githubusercontent.com/134121667/249666682-a10a1a84-321e-484a-ae48-7cd8897190cc.png) ![image](https://user-images.githubusercontent.com/134121667/249666700-02bc6efc-d910-479a-8405-9ee96531c8ee.png)
> > Looking forward to discuss with you.
> 
> We are currently very busy with other ongoing works and try to answer your question in a brief manner. An important characteristic of the SAM is that it can address the ambiguity issue. With a single point, SAM can output three masks to address these issues, for which our MobileSAM follows. If a large object is more meaningful than the smaller one as you suggested, SAM can just choose the three masks that have the largest region. Obviously, this does not make sense and that's why SAM relies on a score ranking mechanism to choose the final mask. In other words, it is the model that automatically prefers the smaller object (without human bias). It seems that FastSAM ignores this ambiguity issue. It seems to me that ignoring this ambiguity issue and claiming a larger object like bus might be biased. After all, for human A, a point can mean the bus, while for human B, the same point can mean the window in this context. Do you agree?

I close it for the moment since there are no follow-up issues. Thanks for your interest in our work again~~

**Comment by Vanessaaui** [comment]
I'm so sorry
